### ISSUE TITLE
Official binaries for `wasm32-unknown-unknown` (and potentially other WASM platforms?) contain code for the wrong architecture

### ISSUE THREAD

**Issue by emilazy** [rust-lang/rust#132802]
The `compiler-builtins` crate compiles C code for WASM platforms since https://github.com/rust-lang/compiler-builtins/pull/566. This works if the C compiler is Clang, as it passes the appropriate `-target`. However, in a GCC build environment this means that the `cc` crate will end up silently compiling code for the wrong architecture entirely. This means that, for example, the `compiler-builtins` shipping for `wasm32-unknown-unknown` via Rustup contains object files like the following:

```
45c91108d938afe8-cmpti2.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), with debug_info, not stripped
```

I suppose that the build for these needs to arrange for Clang to be present, or perhaps even specified explicitly in the `target.*.{cc,cxx,linker}` settings.

(Was redirected here from https://github.com/rust-lang/compiler-builtins/issues/732.)

**Comment by jyn514** [comment]
@emilazy are you interested in fixing this? i think there are three main steps:
1. give a hard error in bootstrap when using gcc to compile for wasm
2. change our CI to use clang instead of gcc
3. add a test that compiling a sample program for wasm32-unknown doesn't give any linker warnings

optionally, you could also change bootstrap to default to `clang` instead of `cc` for wasm32-unknown.

i'm happy to mentor this.

**Comment by JayAndJef** [comment]
Hi @jyn514! If @emilazy doesn't want to take on this issue, I would be glad to. I'm not sure if this is beyond the scope of mine or yours, however, as i'm looking for a good first issue in rustc to resolve - would you be willing to mentor?

**Comment by jyn514** [comment]
I'm still willing to mentor. I think this would be a challenging first issue but you could do it if you're motivated. There are lots of small things to change, but you will (mostly) have short feedback loops, and finding which parts of the code are relevant will not be too hard I think. Take a look at `src/bootstrap`, at the [dev guide for bootstrap](https://rustc-dev-guide.rust-lang.org/building/bootstrapping/what-bootstrapping-does.html), and at the [wasm32-unknown platform docs](https://doc.rust-lang.org/rustc/platform-support/wasm32-unknown-unknown.html#building-the-target). This check should probably (?) live in `sanity.rs`, which [has flaws](https://github.com/rust-lang/rust/issues/136759#issuecomment-2646056717), but it's what we have right now.

**Comment by JayAndJef** [comment]
I think I'll take this issue on. Are there any steps I need to take besides forking the repository, such as claiming this issue or moving to a different platform to communicate with you?

**Comment by jyn514** [comment]
both those, yeah - i think `@rustbot claim` is the current syntax?

i'm on [zulip](https://rust-lang.zulipchat.com/) as "jyn".

**Comment by JayAndJef** [comment]
Great, I'll scope out the bootstrap module

@rustbot claim

**Comment by tgross35** [comment]
> This means that, for example, the `compiler-builtins` shipping for `wasm32-unknown-unknown` via Rustup contains object files like the following:
> 
> ```
> 45c91108d938afe8-cmpti2.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), with debug_info, not stripped
> ```

That seems like the sort of thing that would have gotten noticed almost immediately, compiler-builtins is always provided at linktime. How is this working currently - are builtins never actually needed or is it somehow possible to link wasm and x86 objects together?

**Comment by jyn514** [comment]
> That seems like the sort of thing that would have gotten noticed almost immediately, compiler-builtins is always provided at linktime. How is this working currently - are builtins never actually needed or is it somehow possible to link wasm and x86 objects together?

bootstrap supports disabling `optimized-compiler-builtins` (explicitly, since https://github.com/rust-lang/rust/pull/102579). in that case it uses the rust symbols defined in https://github.com/rust-lang/compiler-builtins/ directly, not the ones defined in LLVM.

you can see in https://github.com/rust-lang/rust/issues/136096 that LLD unconditionally warns when compiling to wasm32-unknown, but it is silenced by rustc so you can't see the warning. I suspect what has happened is:
- someone turned on `optimized-compiler-builtins = true` for wasm32-unknown in rl/r CI and just assumed that would work.
- that had absolutely no effect, because of this bug. instead it fell back to using the weak symbols defined in rust, not the strong symbols defined in llvm.
- because rustc silenced the warning, no one noticed until Emily actually inspected the final binary and noticed that it wasn't using the optimized builtins.

> cc provides [is_like_clang](https://docs.rs/cc/latest/cc/struct.Tool.html#method.is_like_clang). Would things work better if we just disabled the C build if the target is wasm and not is_like_clang?

@tgross35 i think doing that in compiler-builtins would be a good step, yeah. but it still requires work downstream in rust-lang/rust - all the steps in https://github.com/rust-lang/rust/issues/132802#issuecomment-2648125986 are still required; *maybe* we could get away with delaying step 1 until bootstrap actually builds compiler-builtins, but it would ideally be nice to catch it early.

**Comment by jyn514** [comment]
cc https://github.com/rust-lang/rust/issues/105065

**Comment by trevyn** [comment]
> That seems like the sort of thing that would have gotten noticed almost immediately, compiler-builtins is always provided at linktime. How is this working currently - are builtins never actually needed or is it somehow possible to link wasm and x86 objects together?

I think this is a result of https://github.com/rust-lang/compiler-builtins/pull/566

Basically, this doesn’t affect any builtins that are only called by Rust, so if you’re only building Rust code for wasm, you’ll never run into this.

**Comment by jyn514** [comment]
> Basically, this doesn’t affect any builtins that are only called by Rust, so if you’re only building Rust code for wasm, you’ll never run into this.

i am confused - i thought the builtins have the same mangled names at runtime, regardless of whether you use the C builtins or the rust builtins, the only difference is that the rust builtins are weak symbols and the c builtins are strong.

looking at https://github.com/rust-lang/compiler-builtins/pull/566, it seems it doesn't make any changes other than to compile the C builtins? so it makes sense to me it's falling back to the rust ones. but i don't understand what "only called by rust" has to do with anything.

**Comment by trevyn** [comment]
Sorry, I added a spurious “only” there and fixed it in edit.

My understanding is that the only builtins that are miscompiled are ones that Rust never calls.

But if you use clang to compile some C to wasm (to ultimately link with your Rust), it may use these extra builtins.

So it might be useful to make these builtins available.

Which it sounds like you have a plan to make possible. :)

**Comment by HamidrezaSK** [comment]
Hi @jyn514, is this issue still ongoing? I can see there's already a PR open, but it looks like CI is failing, and there have not been any updates since mid-March. I tried reproducing the issue, but wasn’t able to. If the problem is still relevant and needs to be addressed, I’d be happy to take it over.

**Comment by tgross35** [comment]
I think we want the change at https://github.com/rust-lang/rust/pull/137457 even if this somehow fixed itself; since gcc doesn't support wasm, there isn't any reason we shouldn't be erroring out if it somehow winds up used. I just pinged the author of that PR, if they don't respond in a week or so then you can probably pick it up @HamidrezaSK (of course leaving the original author as a co-author).

**Comment by james-rms** [comment]
Since this has been deduped with https://github.com/rust-lang/rust/issues/136109, I want to point out that I can repro this with just clang:
```bash
mkdir mylib
cd mylib
cargo init --lib .
cat << EOF > Cargo.toml
[package]
name = "mylib"
version = "0.1.0"
edition = "2024"

[lib]
name = "mylib"
crate-type = ["cdylib"]
EOF
cargo build --release --target=wasm32-unknown-unknown
objdump -a target/wasm32-unknown-unknown/release/libmylib.a
```

produces similar output when building on my ARM64 Mac. 
`cargo --version`: cargo 1.89.0 (c24e10642 2025-06-23)
`rustc --version`: rustc 1.89.0 (29483883e 2025-08-04)
